### PR TITLE
49 add callSC to the thyra account class and update callSC in the account class

### LINKED ISSUE #49: Add callSC to the thyraAccount class + update callSC  in the Account class

### PR DESCRIPTION
**Pull request by Elli610**
closes #49 

(dont forget to use http instead of https for local testing)

### CODE REVIEW

**Review by julienbrs — APPROVED** [review]

**Review by Ben-Rey — APPROVED** [review]

File: test/simple-browser-extension/src/interfaces/IAccountCallSCRequest.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+/**
+ * This interface represents the request object to send to the content script to 
+ * interact with a deployed smart contract.
+ */
+export interface IAccountCallSCRequest {
+    nickname: string, // account nickname
+    name: string, // function name
+    at: string, // contract address
+    args: Args, // function arguments
+    coins: number, // coins to send to the block creator
```
Comment (gregLibert) [inline test/simple-browser-extension/src/interfaces/IAccountCallSCRequest.ts:12]: coins are normaly expressed in bigint, no ?

**Review by gregLibert — CHANGES_REQUESTED** [review]
Why not planing a dry-run as requested here: https://github.com/massalabs/massa-standards/pull/48#discussion_r1207026899 ?

**Review by gregLibert — CHANGES_REQUESTED** [review]
Why not planing a dry-run as requested here: https://github.com/massalabs/massa-standards/pull/48#discussion_r1207026899 ?

**Comment by Elli610** [comment]
> Why not planing a dry-run as requested here: [massalabs/massa-standards#48 (comment)](https://github.com/massalabs/massa-standards/pull/48#discussion_r1207026899) ?

I just added it

File: test/simple-browser-extension/src/interfaces/IAccountCallSCRequest.ts (hunk 1 above)
Comment (gregLibert) [reply test/simple-browser-extension/src/interfaces/IAccountCallSCRequest.ts:12]: Here the interface is used for massastation but also the browser extension. So the code needs to be compatible for both.
This being said, are you sure a u64 can be expressed using JSON ?

Comment (Elli610) [reply test/simple-browser-extension/src/interfaces/IAccountCallSCRequest.ts:12]: My bad :
"the JSON standard does not have a built-in numeric type that corresponds to u64 (unsigned 64-bit integer). However, you can still include a u64 value in a JSON object without converting it to a string"

Comment (Elli610) [reply test/simple-browser-extension/src/interfaces/IAccountCallSCRequest.ts:12]: Now I have a "Error: Axios error: TypeError: Do not know how to serialize a BigInt" while calling callSC

Comment (Elli610) [reply test/simple-browser-extension/src/interfaces/IAccountCallSCRequest.ts:12]: You are right but axios won't allow it natively. I'll try to make it work

File: src/account/IAccountCallSCRequest.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * @remarks
+ * - If your smart contract function does not require any parameters, you can pass an empty array.
+ *
+ * @see nickname - The nickname of the account to use.
+ * @see name - The name of the function to be called.
+ * @see at - The address of the smart contract.
+ * @see args - The parameters as an Args object to be passed to the function.
+ * @see coins - The amount of MASSA coins to be sent to the block creator.
+ */
```
Comment (Ben-Rey) [inline src/account/IAccountCallSCRequest.ts:16]: You can add dryRun in the code doc

File: src/thyra/ThyraAccount.ts [hunk 3]
```diff
@@ -217,4 +229,138 @@ export class ThyraAccount implements IAccount {
...
+   * @param parameter - The parameters as an Args object to be passed to the function.
+   * @param amount - The amount of MASSA coins to be sent to the block creator.
+   * @param dryRun - The dryRun object to be passed to the function.
+   *
+   * @returns if dryRun.dryRun is true, it returns an IContractReadOperationResponse object. Otherwise,
+   * it returns an ITransactionDetails object which contains the first event emitted by the contract.
+   * If the contract does not emit any event, it returns "Function called successfully but no event generated"
+   *
+   */
+  public async callSC(
```
Comment (Ben-Rey) [inline src/thyra/ThyraAccount.ts:252]: This function do a lot of things, can we split it with small functions. It would be easier to read

**Review by Ben-Rey — CHANGES_REQUESTED** [review]

File: src/thyra/ThyraAccount.ts (hunk 3 above)
Comment (Elli610) [reply src/thyra/ThyraAccount.ts:252]: is the new version ok for you ?

**Review by Ben-Rey — CHANGES_REQUESTED** [review]

File: src/account/IDryRunData.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+/**
+ * This interface defines specific data that are used to perform a dry run.
+ *
+ * @see dryRun - Wether or not to perform a dry run
+ * @see maxGas - The maximum amount of gas to be used in the dry run
+ */
+export interface IDryRunData {
+  dryRun: boolean;
+  maxGas?: bigint;
+}
```
Comment (gregLibert) [inline src/account/IDryRunData.ts:10]:
```suggestion
export interface NonPersistentExecution {
  isNPE: boolean;
  maxGas?: bigint;
}
```

File: src/thyra/ThyraAccount.ts [hunk 5]
```diff
@@ -6,12 +6,24 @@ import {
 } from '..';
 import { IAccount } from '../account/IAccount';
 import { JsonRpcResponseData, getRequest, postRequest } from './RequestHandler';
-import { THYRA_ACCOUNTS_URL } from './ThyraProvider';
+import { THYRA_URL, THYRA_ACCOUNTS_URL } from './ThyraProvider';
```
[the remark this answers is not in the document]
Comment (gregLibert) [reply src/thyra/ThyraAccount.ts:9]: ues should be named MassaStation

**Review by gregLibert — CHANGES_REQUESTED** [review]

**Review by gregLibert — APPROVED** [review]

**Pull request merged**